.github/workflows/build_all.yaml

```diff
@@ -20,95 +20,41 @@ jobs:
-
     - name: Build third party
       if: steps.third_party-cache.outputs.cache-hit != 'true'
       run: nix-shell --run 'cd third_party; make -j1'
 
-    - run: nix-shell --run 'cd cava; make -j1'
     - name: Cache cava
-      uses: actions/cache@v2
-      with:
-        path: |
-          cava
-          /home/runner/.ghc/
-          /home/runner/.cabal/
-        key: ${{ runner.os }}-build-${{ github.run_id }}-cava
-
-  make_demos:
-    needs: make_cava
-    runs-on: ubuntu-latest
-    steps:
-    - uses: actions/checkout@v2
-      with:
-        submodules: 'recursive'
-    - uses: cachix/install-nix-action@v12
-      with:
-        nix_path: nixpkgs=channel:nixos-unstable
-
-
-    - name: Create a "lockfile" from git submodule commits
-      id: get-submodule-ver
-      run: |
-        git submodule status > ./third_party/submodules.lock
-      shell: bash
-    - name: Cache third_party
-      id: third_party-cache
-      uses: actions/cache@v2
-      with:
-        path: third_party
-        key: ${{ runner.os }}-third_party-${{ hashFiles('third_party/submodules.lock') }}
-
-    - name: Use cava cache
-      uses: actions/cache@v2
-      with:
-        path: |
-          cava
-          /home/runner/.ghc/
-          /home/runner/.cabal/
-        key: ${{ runner.os }}-build-${{ github.run_id }}-cava
-    - run: nix-shell --run 'cd demos; make -j1'
-
-  make_examples:
-    needs: make_cava
-    runs-on: ubuntu-latest
-    steps:
-    - uses: actions/checkout@v2
-      with:
-        submodules: 'recursive'
-    - uses: cachix/install-nix-action@v12
-      with:
-        nix_path: nixpkgs=channel:nixos-unstable
-
-    - name: Create a "lockfile" from git submodule commits
-      id: get-submodule-ver
-      run: |
-        git submodule status > ./third_party/submodules.lock
-      shell: bash
-    - name: Cache third_party
-      id: third_party-cache
-      uses: actions/cache@v2
-      with:
-        path: third_party
-        key: ${{ runner.os }}-third_party-${{ hashFiles('third_party/submodules.lock') }}
-
-    - name: Use cava cache
-      uses: actions/cache@v2
-      with:
-        path: |
-          cava
-          /home/runner/.ghc/
-          /home/runner/.cabal/
-        key: ${{ runner.os }}-build-${{ github.run_id }}-cava
-    - run: nix-shell --run 'cd examples; make -j1'
+      id: cava-cache
+      uses: actions/cache@v2
+      with:
+        path: cava
+        key: ${{ runner.os }}-cava-${{ hashFiles('third_party/submodules.lock', 'cava/**/*.v', 'cava/Makefile' ) }}
+    - name: Build Cava
+      if: steps.cava-cache.outputs.cache-hit != 'true'
+      run: nix-shell --run 'cd cava; make -j1'
+
+  make_directories:
+    strategy:
+      matrix:
+        build_dir: ["demos", "examples", "tests"
+        , "silveroak-opentitan/pinmux"
+        , "silveroak-opentitan/aes"
+        , "silveroak-opentitan/hmac"
+        , "examples/xilinx"
+        , "tests/xilinx"
+        , "firmware"
+        ]
```

Comment:
If I understand correctly, `matrix` might run all its task in parallel, in an unspecified order. At the moment, that's fine, but soon, the `firmware` directory will reference `silveroak-opentitan/hmac/Spec/SHA256.v`, so that we can state a theorem saying that the driver function outputs a digest computed according to the definition of `sha256` in that file. At that point, we'll need to make sure that the `firmware` directory is only built once the `silveroak-opentitan/hmac` has finished building.
But since that dependency is not yet present, I'd be also ok with merging this PR as-is.

Reply:
Also, the same problem already exists with `investigations/cava2/Sha256Properties.v`, but since that directory is not run on CI, it's "fine" as well at the moment :wink: